### DISCUSSION TITLE
Getting stabilityDays to work with Python and AWS CodeArtifact

### DISCUSSION THREAD

**Discussion by miikka** [renovatebot/renovate#20470] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.104.0

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

We're using Renovate to upgrade the dependencies in a bunch of Poetry projects. Some but not all projects are configured with `tool.poetry.source` to get the deps from our private AWS CodeArtifact repository which in turn gets public/open-source packages from PyPI.

We would like to use the `stabilityDays` setting to delay the upgrades. This works fine with the projects that use PyPI directly as the source, but it does not seem to work with ones that use CodeArtifact. I'm going to guess that the issue is the [same as with Artifactory](https://github.com/renovatebot/renovate/issues/20070): CodeArtifact only has the "simple" API which does not expose release timestamps. Here's what [CodeArtifact docs](https://docs.aws.amazon.com/codeartifact/latest/ug/python-compatibility.html) say:
> CodeArtifact does not support PyPI's XML-RPC or JSON APIs. CodeArtifact supports PyPI's Legacy APIs, except the simple API. While CodeArtifact does not support the `/simple/` API endpoint, it does support the `/simple/<project>/` endpoint.

Is there some workaround we could use? We do not use Renovate to upgrade the private packages we depend on, so getting the release dates from PyPI would be good enough.

### What you tried so far.

We have set `stabilityDays` to `14`. However, for CodeArtifact-using projects, Renovate proposes upgrades to dependencies as soon as they're released instead of waiting 14 days.

We also tried setting `secondary = true` for our CodeArtifact registry in `pyproject.toml`, but that did not change the situation.

**Comment by viceice** [comment, marked as answer]
if you just mirror public packages, you can override `registryUrls` in renovate config, so renovate searches on pypi instead.

**Reply by miikka** [reply]
Great, thanks, it works! For reference, here's the package rule I added to my config where `secret-prefix-` stands for the package name prefix we use for our internal packages:

```json
{
  "description": "Use PyPI as the registry for public packages so that stabilityDays works",
  "excludePackagePrefixes": ["secret-prefix-"],
  "matchDatasources": ["pypi"],
  "registryUrls": ["https://pypi.org/pypi"]
}
```

**Reply by viceice** [reply]
thanks for sharing the full solution
